### PR TITLE
B2g dev prepare internal api

### PR DESCRIPTION
**Pull request by shane-tomlinson**
@zaach & @jedp - Can you review this one?

This prepares the internal API for B2G. There is a set of tests that go along with this, but those must be merged once we have updated user.js to have "setIssuer"

### CODE REVIEW

File: resources/static/dialog/js/misc/internal_api.js [hunk 1]
```diff
@@ -61,6 +61,13 @@
    * options.silent defaults to false.
    */
   internal.get = function(origin, callback, options) {
+    if (typeof options === 'string') {
+      // Firefox forbids sending objects across the blood-brain barrier from
+      // gecko into userland JS.  So we just stringify and destringify our
+      // objects when calling from b2g native code.
+      options = JSON.parse(options);
```
Comment (jedp) [inline resources/static/dialog/js/misc/internal_api.js]: Maybe throw a `try ... catch` around this and log any weird strings that come from firefox land?

File: resources/static/dialog/js/misc/internal_api.js [hunk 2]
```diff
@@ -163,4 +170,73 @@
     user.logoutUser(callback, complete.curry(null));
   };
 
+  internal.watch = function (callback, options, log) {
+    if (typeof options === 'string') options = JSON.parse(options);
```
Comment (jedp) [inline resources/static/dialog/js/misc/internal_api.js]: Ditto for `try ... catch`, or refactor into a `parseOptions` function that returns object or null?

**Comment by jedp** [comment]
Apart from questions about `try ... catch` around `JSON.parse`, this is looking good to me!

File: resources/static/dialog/js/misc/internal_api.js [hunk 3]
```diff
@@ -163,4 +170,73 @@
...
+
+  function internalWatch (callback, options, log) {
+    var bid = BrowserID,
+        network = bid.Network,
+        user = bid.User,
+        storage = bid.Storage;
+
+    network.init();
+
+    log('internal watch options', options);
```
Comment (shane-tomlinson) [inline resources/static/dialog/js/misc/internal_api.js]: Can we safely remove all of these logs before we merge?

**Comment by shane-tomlinson** [comment]
@jedp - I added parseOptions, which logs the invalid JSON. Should the callback to .watch be called with some sort of error message when this happens?

**Comment by shane-tomlinson** [comment]
@jedp - aside from the logs all over the place, I am comfortable with this. Should we take those out? I have made the changes for parseOptions. What more needs done for this one to get the initial PR in? The updates dealing with forceIssuer I would rather handle in that PR separately.

File: resources/static/dialog/js/misc/internal_api.js [hunk 4]
```diff
@@ -163,4 +186,78 @@
...
+    } catch(e) {
+      // should we call the callback with an error message?
+      return;
+    }
+    internalWatch(callback, options, log);
+  };
+
+
+  function internalWatch (callback, options, log) {
+    var bid = BrowserID,
```
Comment (shane-tomlinson) [inline resources/static/dialog/js/misc/internal_api.js]: The external dependencies should be moved to the top of the file.

File: resources/static/dialog/js/misc/internal_api.js [hunk 5]
```diff
@@ -163,4 +186,78 @@
...
+        doReady();
+      }, function(err) {
+        log('silent return: err', err);
+        doLogout();
+        loggedInUser = null;
+        doReady();
+      }, log);
+    }
+
+    network.clearContext();
```
Comment (shane-tomlinson) [inline resources/static/dialog/js/misc/internal_api.js]: These two function calls live on their own down here and are easy to miss, they should be moved up before checkAndEmit

File: resources/static/dialog/js/misc/internal_api.js [hunk 6]
```diff
@@ -163,4 +186,78 @@
...
+    checkAndEmit();
+
+    function doReady (params) {
```
Comment (shane-tomlinson) [inline resources/static/dialog/js/misc/internal_api.js]: spacing before the (

File: resources/static/dialog/js/misc/internal_api.js [hunk 7]
```diff
@@ -163,4 +186,78 @@
...
+      log('doReady');
+      callback({ method: 'ready' });
+    }
+
+    function doLogin (params) {
```
Comment (shane-tomlinson) [inline resources/static/dialog/js/misc/internal_api.js]: spacing before the (

File: resources/static/dialog/js/misc/internal_api.js [hunk 8]
```diff
@@ -163,4 +186,78 @@
...
+      log('doLogin (with silent assertion)');
+      // Through the _internalParams, we signify to any RP callers that are
+      // interested that this assertion was acquired without user interaction.
+      callback({ method: 'login', assertion: params, _internalParams: {silent: true} });
+    }
+
+    function doLogout () {
```
Comment (shane-tomlinson) [inline resources/static/dialog/js/misc/internal_api.js]: spacing before the (

File: resources/static/dialog/js/misc/internal_api.js [hunk 9]
```diff
@@ -163,4 +186,78 @@
...
+      log('doLogout');
+      if (loggedInUser !== null) {
+        storage.setLoggedIn(remoteOrigin, false);
```
Comment (shane-tomlinson) [inline resources/static/dialog/js/misc/internal_api.js]: This needs to be updated for the storage simplification that occurred. storage.setLoggedIn no longer exists, this is now taken care of by storage.site.set

File: resources/static/dialog/js/misc/internal_api.js [hunk 10]
```diff
@@ -163,4 +187,68 @@
     user.logoutUser(callback, complete.curry(null));
   };
 
+  internal.watch = function (callback, options, externalLog) {
+    log = externalLog || bid.Helpers.log;
+
+    try {
+      options = parseOptions(options);
+    } catch(e) {
+      // should we call the callback with an error message?
```
Comment (jedp) [inline resources/static/dialog/js/misc/internal_api.js]: That would be helpful for the caller.  We callback with an object; in this case, we could `callback({error: e.toString()})` and then throw?

Comment (shane-tomlinson) [inline resources/static/dialog/js/misc/internal_api.js]: I think that would be extremely useful.

**Comment by jedp** [comment]
@shane-tomlinson this is looking good to me.  I made the changes to the `internal.watch` callback you suggested; it now sends the error text if there was a problem parsing the options.

Since this feature is new, I'm comfortable leaving the logs in, with the intention of removing them in the future.

What do you think?  Push the Big Green Button?

**Pull request merged**